### PR TITLE
[SPARK-14400] [SQL] ScriptTransformation does not fail the job for bad user command

### PR DESCRIPTION
**Pull request by tejasapatil**
## What changes were proposed in this pull request?
- Refer to the Jira for the problem: jira : https://issues.apache.org/jira/browse/SPARK-14400
- The fix is to check if the process has exited with a non-zero exit code in `hasNext()`. I have moved this and checking of writer thread exception to a separate method.
## How was this patch tested?
- Ran a job which had incorrect transform script command and saw that the job fails
- Existing unit tests for `ScriptTransformationSuite`. Added a new unit test

### CODE REVIEW

**Comment by tejasapatil** [comment]
ok to test

**Comment by rxin** [comment]
cc @srowen

**Comment by tejasapatil** [comment]
ok to test

**Comment by tejasapatil** [comment]
ok to test

**Comment by tejasapatil** [comment]
ok to test

**Comment by tejasapatil** [comment]
ok to test

**Comment by tejasapatil** [comment]
ok to test

**Comment by tejasapatil** [comment]
Can anyone please review this PR ?

**Comment by tejasapatil** [comment]
Can anyone please review this PR ?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformation.scala [hunk 1]
```diff
@@ -127,45 +127,78 @@ case class ScriptTransformation(
...
+          // The ideal way to do this is to use Java 8's Process#isAlive()
+          // but it cannot be used because Spark still supports Java 7.
+          // Following is a workaround used to check if a process is alive in Java 7
+          // TODO: Once builds are switched to Java 8, this can be changed
+          try {
+            val exitCode = proc.exitValue()
+            if (exitCode != 0) {
+              logError(stderrBuffer.toString) // log the stderr circular buffer
+              throw new SparkException(s"Subprocess exited with status $exitCode. " +
+                s"Error: ${stderrBuffer.toString}", cause)
```
Comment (sameeragarwal) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformation.scala]: unrelated to this PR but we should probably only print the partial contents of the circular buffer if the number of bytes written to it are less than its total size

**Comment by sameeragarwal** [comment]
This patch looks good to me. @srowen / @JoshRosen  can one of you please take a second look as well?

**Comment by rxin** [comment]
@srowen want to review this? Given you reviewed the last transform pr.

**Comment by rxin** [comment]
@tejasapatil on a related note, we'd want to remove as much hive code dependency as possible. One command that is left is ScriptTransformation. Would you have time to implement this in sql/core without Hive dependency for ScriptTransformation? It seems like one of the primary operators you guys would use.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformation.scala [hunk 2]
```diff
@@ -127,45 +127,78 @@ case class ScriptTransformation(
...
+            if (outputSerde == null) {
               if (curLine == null) {
-                if (writerThread.exception.isDefined) {
-                  throw writerThread.exception.get
+                curLine = reader.readLine()
+                if (curLine == null) {
+                  checkFailureAndPropagate()
+                  false
+                } else {
+                  true
```
Comment (srowen) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformation.scala]: Hm, is it me or does this get hard to follow the return values. Generally the method returns "true" unless one of several conditions caused it to decide it was finished earlier. Those could be handled with early "return false" rather than lots of "else ... true" branches.

**Comment by srowen** [comment]
With one minor suggestion this LGTM

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformation.scala (hunk 2 above)
Comment (rxin) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformation.scala]: yea i think that's a great idea. the level of nesting is a little bit too much here

**Comment by rxin** [comment]
@tejasapatil want to update this so we can merge it for 2.0?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformation.scala (hunk 1 above)
Comment (tejasapatil) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformation.scala]: @sameeragarwal : Thanks for pointing that out. I will submit a separate PR for that.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformation.scala (hunk 2 above)
Comment (tejasapatil) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformation.scala]: +1. Did the change

**Comment by tejasapatil** [comment]
@rxin : Sorry for delay... was caught up in some other things. I have updated the PR now with the review comments. 

Re your suggestion about removing hive code dependency: I will work on it

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformation.scala (hunk 1 above)
Comment (tejasapatil) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformation.scala]: here : https://github.com/apache/spark/pull/13351

**Comment by sameeragarwal** [comment]
LGTM pending jenkins. Thanks!

**Comment by rxin** [comment]
Merging in master/2.0.
